bin/problem.py

```diff
@@ -172,17 +172,73 @@ def _read_settings(self):
                 stored_uuid.write_text(self.settings.uuid)
             warn(f'Missing UUID for {self.name}, add to problem.yaml:\nuuid: {self.settings.uuid}')
 
-    # Walk up from absolute `path` (a file or directory) looking for the first testdata.yaml
-    # file, and return its contents, or None if no testdata.yaml is found.
-    def get_testdata_yaml(p, path):
+    def get_testdata_yaml(p, path, key, name=None, bar=None) -> str | None:
+        """
+        Find the testdata flags applying at the given path for the given key.
+        If necessary, walk up from `path` looking for the first testdata.yaml file that applies,
+
+        Side effects: parses and caches the file.
+
+        Arguments
+        ---------
+        path: absolute path (a file or a directory)
+        key: The testdata.yaml key to look for, either of 'input_validator_flags', 'output_validator_flags', or 'grading'.
+            'grading' is not implemented
+        name: If key == 'input_validator_flags', optionally the name of the input validator
+
+        Returns:
+        --------
+        string or None if no testdata.yaml is found.
+        TODO: when 'grading' is supported, it also can return dict
+        """
+        if key == 'grading':
+            raise NotImplementedError(key)
+        if key != 'input_validator_flags' and name is not None:
+            raise ValueError(
+                f"Only input validators support flags by validator name, got {key} and {name}"
+            )
+
+        error_handler = error if bar is None else bar.error
+        warning_handler = warn if bar is None else bar.warn
+
         for dir in [path] + list(path.parents):
             f = dir / 'testdata.yaml'
 
             if f.is_file():
-                # Store testdata.yaml files in a cache.
-                if f not in p._testdata_yamls:
-                    p._testdata_yamls[f] = read_yaml(f, plain=True)
-                return p._testdata_yamls[f]
+                if f in p._testdata_yamls:
+                    flags = p._testdata_yamls[f]
+                else:
+                    p._testdata_yamls[f] = flags = read_yaml(f, plain=True)
+
+                # Validate and exctract the flags
+                for k in flags:
+                    match k:
+                        case 'output_validator_flags':
+                            if not isinstance(flags[k], str):
+                                error_handler("ouput_validator_flags must be string")
```

Comment:
This is slightly inconsistent:
`output_validator_flags` is checked for `str` only if `output_validator_flags` comes before `input_validator_flags` in case `key == input_validator_flags`.
I'd say either:
- First loop to check that all keys are known and correct style, and then return `flags[k]`.
- Only check `flags[k]`.

Reply:
… you mean “only check `flags[key]`, right? (I.e., the method argument, rather than the iteration variable.)

Reply:
yes, indeed